Wifi scanning protocol extension

Tomcus opened the pull request:
Extending the esp protocol to allow printer to ask esp to start scanning for available wifi APs.

BFW-5387

Tomcus committed:
Make rx_callback take whole header instead of size

This should allow more flexibility in future commits when sending data
from esp to printer.

BFW-5387

Tomcus committed:
Make wifi scanning more generic

Making wifi scanning more generic to make it extandable in the future.

BFW-5387

Tomcus committed:
Extend the protocol to be able to do asynchronous wifi scan

This change allows printer to ask esp to start scanning for available
networks. The results are stored on ESP side to preserve memory on the
printer. The results can be retrieved from the ESP over uart and the ESP
periodically updates the printer about number of found access points.

BFW-5387

vorner commented on main/uart_nic.c:
```diff
@@ -133,12 +143,42 @@ static uint32_t IRAM_ATTR now_seconds() {
+    wifi_scan_callback callback;
+    ScanResult stored_ssids[SCAN_MAX_STORED_SSIDS];
+    uint8_t stored_ssids_count;
+    volatile bool in_progress;
```
The fact we need `volatile` scares me. Do you, by any chance, mean `atomic` instead?

vorner commented on main/uart_nic.c:
```diff
@@ -151,24 +191,106 @@ static void IRAM_ATTR send_link_status(uint8_t up) {
-static void IRAM_ATTR probe_task(void* arg) {
-    wifi_scan_config_t config;
+static void IRAM_ATTR do_wifi_scan(void * arg) {
+    wifi_scan_config_t config = {0};
```
Actually, zeroes are implicit (at least in the C version), so to set something to all-zeroes, one can do just `= {}`.

(Multiple instances through the file)

Tomcus replied:
I know I just prefer to be explicit.

vorner replied:
Looks a bit weird, as the struct has more than one zeroable field 😇. But OK, I'll leave it up to you.

vorner commented on main/uart_nic.c:
```diff
@@ -403,37 +514,115 @@ static void IRAM_ATTR check_online_status() {
 
+static void IRAM_ATTR clear_stored_ssids() {
+    memset(scan.stored_ssids, 0, scan.stored_ssids_count * sizeof(ScanResult));
+    scan.stored_ssids_count = 0;
```
Is it enough to just set the count to 0?

vorner commented on main/uart_nic.c:
```diff
@@ -403,37 +514,115 @@ static void IRAM_ATTR check_online_status() {
+    for (uint8_t i = 0; i < ap_count; ++i) {
+        bool found = false;
+        if (scan.stored_ssids_count == SCAN_MAX_STORED_SSIDS) {
+            break;
```
Would this thing deserve its own log message too?

vorner commented on main/uart_nic.c:
```diff
@@ -46,7 +46,11 @@
-#define FW_VERSION 11
+#define FW_VERSION 12
+
+#define SCAN_MAX_STORED_SSIDS 64
```
Is this based on anything, or just „randomly chosen number“? Considering we likely have enough RAM on the ESP side, it could make sense to just overestimate by a bit (I can imagine someone seeing 64 or more SSIDs in eg. a big city somewhere).

Tomcus replied:
Since the terrible results in the office, I don't think the small esp with crappy antenna will ever find more than few AP around. Of course I can increase it if needed. I can put 128, but would love to test it more if it doesn't break anything.

vorner commented on main/uart_nic.c:
```diff
@@ -403,37 +514,115 @@ static void IRAM_ATTR check_online_status() {
+            }
+        }
+        if (!found) {
+            memcpy(scan.stored_ssids[scan.stored_ssids_count].ssid, aps[i].ssid, SSID_LEN);
```
:thinking: Is memcpy the right one? Is aps[i].ssid also an array of at least such size, or is it just a null-terminated string and the valid memory might be shorter?

Tomcus replied:
Yes it is `uint8_t ssid[33]`. https://docs.espressif.com/projects/esp8266-rtos-sdk/en/latest/api-reference/wifi/esp_wifi.html#_CPPv4N16wifi_ap_record_t4ssidE

vorner replied:
Then maybe a static_assert checking that both things have the same size, eg `static_assert(sizeof(stored_ssids.ssid) == sizeof(aps.ssid) && == SSID_LEN)`? Just to make sure nothing breaks terribly if we ever update the ESP SDK or something.

Tomcus replied:
Sadly I can't use `static_assert` in the c codebase. We should be able to use c++ and might be a good idea to investigate it in the future.

vorner commented on main/uart_nic.c:
```diff
@@ -403,37 +514,115 @@ static void IRAM_ATTR check_online_status() {
+}
+
+static void IRAM_ATTR handle_rx_msg_scan_get(uint8_t* data, struct header header) {
+    send_scanned_ssid(header.ap_index);
```
Should there be some kind of „range check“ somewhere?

Tomcus replied:
I would prefer that it would happen on the printer side since the protocol doesn't support sending errors. I can modulo the index (but need to zero out the data that you object about to be able to return zeroth element if none are found) or I can "silently" fail (I can log an error, but nobody will know about it until they connect to the error pin with something that can read it).

vorner replied:
In these areas, I prefer to be somewhat defensive. I fully agree the primary check should be on the printer side and I'm fine classifying such event protocol violation / printer fault / bug.

I'm merely proposing to limit the „blast radius“ of such bug ‒ silently not answering, or providing some kind of null/empty answer instead of crashing the ESP firmware or sending some (potentially secret data) over the wire.

In a sense, it's a separate firmware running on a separate chip and it should do some (at least minimal) diligence against external communications ‒ that is, not expect whatever comes to be correct and non-malicious.

Anyway, any kind of minimal check and handling would be sufficient.

vorner commented on protocol.txt:
```diff
@@ -123,3 +127,93 @@ MSG_PACKET_V2
+       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+    Sends a signal to esp to start a wifi scan. The scan is iterative
+    and result are sent back to the printer periodically. The first
```
The `results are sent back periodically` would hint at the ESP sending the list of APs from time to time. Is that really the case? I thought the printer is „asking“ for nth SSID when it needs it.

vorner commented on protocol.txt:
```diff
@@ -123,3 +127,93 @@ MSG_PACKET_V2
+
+    AP_INDEX    This value is only send to the ESP by the printer and
+                is not set in the response. This values should be valid
+                index into the internal ap storage
```
I'd suggest we „echo back“ the index, so the printer can check it.

vorner commented on protocol.txt:
```diff
@@ -123,3 +127,93 @@ MSG_PACKET_V2
+                index into the internal ap storage
+
+    AP_INFO_LEN Length of send ap info data. The value is fixed in
+                every version (but might change based on the version).
```
What units? Bytes?

Must be 0 when sent by printer?

vorner commented on protocol.txt:
```diff
@@ -123,3 +127,93 @@ MSG_PACKET_V2
+
+    AP_INFO     Binary representation of the stored ap info in ESP.
+                For more info check the ScanData structure in
+                uart_nic.c.
```
Only present when sent by the ESP?

You mean ScanResult? (There's no ScanData there).

Btw, if sent on the wire, it should at least be „packed“. Even though there are only 1-byte fields, the compiler is free to add padding and we are not 100% sure it would do the same padding on both sides (because potentially different compilers).

Tomcus committed:
fix: Send correct data from wifi

Tomcus committed:
fix: Delay handeling of wifi disconnect after a scan

This should fix resets during long scans.

BFW-5387

Tomcus committed:
fix: Scan timings

The esp scan behaves differently when connected to AP and disconnected
from AP. To prevent this issue we intentionally disconnect from the wifi
if connected and use the delayed disconnect handling (done in previous
commit) to reconnect after scan.

Also changed the waiting times of the scan to be more incremental.

These changes makes the results appear faster with more incremental
popins.

BFW-5387

Tomcus committed:
scan: Fix repoting false possitive networks without password

Authored by: Patrik Modesto <mody@centrum.cz>

BFW-5387

Tomcus committed:
probe: fix crashing when getting ap_info with disconnected wifi

This should fix resets when wifi disconnects and we try to tell if the
AP is available. Now we cache the bssid and ssid when we connect so we
don't have to inquire connected ap info when the wifi is disconnected.

Tomcus committed:
wifi: fix password handeling issue

Since the password isn't guaranteed to be null terminated so we can't
call strlen on it.

Authored by: Patrik Modesto <mody@centrum.cz>

Tomcus committed:
Fix connect during wifi scan

If we want to connect to new wifi when scanning we need to stop the
scan and try to prevent reconnects since we are connecting to new
network. This decreases the retries to connect to APs.

BFW-5579

Tomcus committed:
fix: Prevent common disconnects

BFW-5640

Tomcus committed:
Don't include source file, include header file

Tomcus committed:
esp: Run connecting in separate task

By making a wifi connection non blocking we are able to start the wifi
scan. Also there is need for some small synchronization to not start
wifi scan during connecting (state change).

BFW-5633

Tomcus commented:
Closing in favor of #3 with better named branch
